drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1454 @@
+static KSPIN_LOCK SocketsToPutListLock = 0;
+static BOOLEAN PutSocketsThreadShouldRun = FALSE;
+static KEVENT PutSocketsEvent;
+static HANDLE PutSocketsThreadHandle;
+
+static void WSKAPI PutSocketsThread(void *p)
+{
+    NTSTATUS status;
+    PWSK_SOCKET_INTERNAL SocketToPut;
+    KIRQL flags;
+    int NumSocketPuts;
+
+    FUNCTION_TRACE;
+
+    while (PutSocketsThreadShouldRun)
+    {
+        status = KeWaitForSingleObject(&PutSocketsEvent, Executive, KernelMode, FALSE, NULL);
+        if (!NT_SUCCESS(status))
+        {
+            DbgPrint("KeWaitForSingleObject failed with status 0x%08x!\n", status);
+        }
+
+        KeAcquireSpinLock(&SocketsToPutListLock, &flags);
+        while (SocketsToPut != NULL)
+        {
+            SocketToPut = SocketsToPut;
+            SocketsToPut = SocketToPut->NextSocketToPut;
+            NumSocketPuts = SocketToPut->NumSocketPuts;
+
+            KeReleaseSpinLock(&SocketsToPutListLock, flags);
+            while (NumSocketPuts > 0)
+            {
+                SocketToPut->NumSocketPuts--;
+                NumSocketPuts = SocketToPut->NumSocketPuts;
+
+                SocketPut(SocketToPut);	/* Here IRQL MUST be PASSIVE_LEVEL, else we loop forever! */
+            }
+            KeAcquireSpinLock(&SocketsToPutListLock, &flags);
+        }
+        KeReleaseSpinLock(&SocketsToPutListLock, flags);
+    }
+}
+
+static void StartSocketPutThread(void)
+{
+    NTSTATUS status;
+
+    KeInitializeSpinLock(&SocketsToPutListLock);
+    KeInitializeEvent(&PutSocketsEvent, SynchronizationEvent, FALSE);
+    PutSocketsThreadShouldRun = TRUE;
+
+    status = PsCreateSystemThread(&PutSocketsThreadHandle, THREAD_ALL_ACCESS, NULL, NULL, NULL, PutSocketsThread, NULL);
+    if (status != STATUS_SUCCESS)
+    {
+        DbgPrint("Could not start put sockets thread, status is %x\n", status);
+    }
+}
+
+static void StopSocketPutThread(void)
+{
+    PutSocketsThreadShouldRun = FALSE;
+    KeSetEvent(&PutSocketsEvent, IO_NO_INCREMENT, FALSE);
+
+    /* eventually it will terminate, no need to wait for that. */
+}
+
+void
+SocketPut(PWSK_SOCKET_INTERNAL s)
+{
+    KIRQL flags;
+
+    FUNCTION_TRACE;
+
+    /* We are doing calls like ZwClose, TdiDisassociateAddressFile, ...
+       which require IRQL to be PASSIVE_LEVEL. So if we get here while
+       in IRQ, defer the SocketPut() to a separate thread. */
+
+    if (KeGetCurrentIrql() > PASSIVE_LEVEL)
+    {
+        KeAcquireSpinLock(&SocketsToPutListLock, &flags);
+        if (s->NumSocketPuts > 0)
+        {
+            s->NumSocketPuts++;
+        }
+        else
+        {
+            s->NextSocketToPut = SocketsToPut;
+            SocketsToPut = s;
+            s->NumSocketPuts = 1;
+        }
+        KeReleaseSpinLock(&SocketsToPutListLock, flags);
+
+        KeSetEvent(&PutSocketsEvent, IO_NO_INCREMENT, FALSE);
+
+        return;
+    }
+
+    s->RefCount--;
+    if (s->RefCount == 0)
+    {
+        SocketShutdown(s);	/* noop when called twice */
```

Comment:
If this comment is general to the SocketShutdown function, I would suggest you move it to the SocketShutdown function itself as part of its documentation.